working/augmentation-libraries/parts_with_imports.md

```diff
@@ -312,18 +414,93 @@ Also, every file in a library must belong to the same *package*. The Dart
 language itself has no notion of packages, but the tooling uses a file’s
 package to derive its default language version. The Dart SDK will require that
 all files in a package belong to the same library, ensuring that they’ll always
-have the same language version. We haven’t specified this requirement before,
+have the same language version. Also, if the library file is inside the `lib/`
+directory, so it has a `package:` URI as canonical URI, then so must all its
+sub-part files. We haven’t specified these requirements before,
 it has always been assumed, but technically it is possible to write programs
 where a part belongs to different package than their library. The Dart SDK’s
 multi-language-version support, which based on files belonging to packages,
 will not support libraries that are not entirely in a single package.
 
+That is, the extra restrictions enforced by Dart tools are:
+
+*   If any file of a library has a `// @dart=` language version marker, then
+    *all* files in that library must have a language version marker with
+    *the same* language version.
+    Or, phrased differently starting at the root:
+    *   If a library file has a language version marker,
+        then it’s a compile-time error for any sub-part file
+        to not have the same language version marker.
+    *   If a library file has no language version marker,
```

Comment:
Ok hear me out - what if you have a part which is defined in a different package, with a different default language version.

Should we explicitly block parts from other packages? Generally the spec doesn't talk about packages, but it seems like a reasonable constraint.

If we do allow it, then I think this should be specified not in terms of the existence of a language version marker, but in terms of the actual computed language version, which must be the same across all parts and the root library file.

Reply:
> Ok hear me out - what if you have a part which is defined in a different package, with a different default language version. Then you are required to have a language marker in those parts.

And that's the reason I explicitly specified that as an error. :sunglasses: 

> Should we explicitly block parts from other packages? 

We should, and I did! 
Part files must belong to the same package as their parent file. That ensures they have the same *default* language version.

We always should have said that, precisely because of this issue.
We've said that a library file and its part file must have the same language file marker, or none.
What we really mean is that they must have the same language version, but we also want that to stay true if someone changes the SDK version.

The *best way*<sup>&tm;</sup> to ensure that they have the same version *no matter what else changes* is to require either all having the same language version marker override, or all having the same default version, and the latter is implied by being in the same package.